massa-models/src/denunciation.rs

```diff
@@ -0,0 +1,642 @@
+                    &endo_bh.public_key,
+                    &endo_bh.slot,
+                    &content_hash,
+                );
+
+                if let Ok(hash) = hash_ {
+                    Ok(endo_bh.slot == s_block_header.content.slot
+                        && endo_bh.public_key == s_block_header.content_creator_pub_key
+                        && endo_bh.hash_1 != content_hash
+                        && endo_bh.hash_2 != content_hash
+                        && endo_bh
+                            .public_key
+                            .verify_signature(&hash, &s_block_header.signature)
+                            .is_ok())
+                } else {
+                    Ok(false)
+                }
+            }
+        }
+    }
+
+    fn is_valid(&self) -> Result<bool, DenunciationError> {
+        let (signature_1, signature_2, hash_1, hash_2, public_key) = match self {
+            Denunciation::Endorsement(de) => {
+                let hash_1 = EndorsementDenunciation::compute_hash_for_sig_verif(
+                    &de.public_key,
+                    &de.slot,
+                    &de.index,
+                    &de.hash_1,
+                )?;
+                let hash_2 = EndorsementDenunciation::compute_hash_for_sig_verif(
+                    &de.public_key,
+                    &de.slot,
+                    &de.index,
+                    &de.hash_2,
+                )?;
+
+                (
+                    de.signature_1,
+                    de.signature_2,
+                    hash_1,
+                    hash_2,
+                    de.public_key,
+                )
+            }
+            Denunciation::BlockHeader(de) => {
+                let hash_1 = BlockHeaderDenunciation::compute_hash_for_sig_verif(
+                    &de.public_key,
+                    &de.slot,
+                    &de.hash_1,
+                )?;
+                let hash_2 = BlockHeaderDenunciation::compute_hash_for_sig_verif(
+                    &de.public_key,
+                    &de.slot,
+                    &de.hash_2,
+                )?;
+
+                (
+                    de.signature_1,
+                    de.signature_2,
+                    hash_1,
+                    hash_2,
+                    de.public_key,
+                )
+            }
+        };
+
+        Ok(hash_1 != hash_2
+            && signature_1 != signature_2
+            && public_key.verify_signature(&hash_1, &signature_1).is_ok()
+            && public_key.verify_signature(&hash_2, &signature_2).is_ok())
+    }
+}
+
+/// Create a new Denunciation from 2 SecureShareEndorsement
+impl TryFrom<(&SecureShareEndorsement, &SecureShareEndorsement)> for Denunciation {
+    type Error = DenunciationError;
+
+    fn try_from(
+        (s_e1, s_e2): (&SecureShareEndorsement, &SecureShareEndorsement),
+    ) -> Result<Self, Self::Error> {
+        // Cannot use the same endorsement twice
+        if s_e1 == s_e2 {
```

Comment:
we can compare IDs

Reply:
I don't remember, maybe this is only useful for tests ?

Reply:
Add a TODO so this will be done in a follow up